spec/forms/user_form_spec.rb

```diff
@@ -12,41 +12,107 @@
-  end
+    it "if user is scientist then they can edit themselves" do
+      expect {
+        UserForm.new(scientist).submit(params.merge(user: { swipe_card_id: '12345', user_code: sci_swipe_card_id }))
+      }.to change { scientist.reload.swipe_card_id }
+    end
 
-  it "if user is authorised but swipe card id is left blank then swipe card id should not be updated" do
-    expect {
-      subject.submit(params.merge(user: { swipe_card_id: nil, user_code: admin_swipe_card_id }, commit: "Update User"))
-    }.to_not change { technician.reload.swipe_card_id }
-  end
+    it "if user is scientist then they can edit themselves but not change their type" do
+      scientist_form = UserForm.new(scientist)
+      expect {
+        scientist_form.submit(params.merge(user: { type: "Administrator", user_code: sci_swipe_card_id }))
+      }.to_not change { scientist.reload.type }
+      expect(scientist_form.errors.full_messages).to include("You do not have permission to change your user type")
+    end
 
-  it "if user is authorised but barcode is left blank then barcode should not be updated" do
-    expect {
-      subject.submit(params.merge(user: { barcode: nil, user_code: admin_swipe_card_id }, commit: "Update User"))
-    }.to_not change { technician.reload.barcode }
-  end
+    it "if user is authorised and swipe card is valid it should be updated" do
+      expect {
+        subject.submit(params.merge(user: { swipe_card_id: '12345', user_code: admin_swipe_card_id }))
+      }.to change { technician.reload.swipe_card_id }.to(Digest::SHA1.hexdigest('12345'))
+    end
+
+    it "if user is authorised but swipe card id is left blank then swipe card id should not be updated" do
+      expect {
+        subject.submit(params.merge(user: { swipe_card_id: nil, user_code: admin_swipe_card_id }))
+      }.to_not change { technician.reload.swipe_card_id }
+    end
+
+    it "if user is authorised but barcode is left blank then barcode should not be updated" do
+      expect {
+        subject.submit(params.merge(user: { barcode: nil, user_code: admin_swipe_card_id }))
+      }.to_not change { technician.reload.barcode }
+    end
+
+    it "if user is authorised they should be able to update a users team" do
+      expect {
+        subject.submit(params.merge(user: { team_id: 1, user_code: admin_swipe_card_id }))
+      }.to change { technician.reload.team_id }.to(1)
+    end
+
+    it "if no values are entered a user should not change" do
+      expect {
+        subject.submit(params.merge(user: { user_code: admin_swipe_card_id }))
+      }.to_not change { technician.reload }
+    end
 
-  it "if user is authorised they should be able to update a users team" do
-    expect {
-      subject.submit(params.merge(user: { team_id: 1, user_code: admin_swipe_card_id }, commit: "Update User"))
-    }.to change { technician.reload.team_id }.to(1)
+    it "if user is technician they should be able to update a non-admin user" do
```

Comment:
I guess here there is a technician updating a technician so it might be testing another bit of code, so maybe change it to a technician updating a scientist